### PR TITLE
[RNMobile] refactor InspectorControls

### PR DESCRIPTION
**Pull request by jbinda**
## Description
PR is connected with [#1300](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1300) in gutenberg-mobile.

Please also refer to [Related gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1341)

It presents:
1. PanelBody mobile implementation (currently simple `View` component)
2. Refactor of `InspectorControls` that allows to use it in pretty same manner as on web
3. Refactor in `edit.native.js` in image and video block
4. Update `SettingsButton` in `BlockMobileToolbar` behaviour to allow show/hide depending if there is any settings available to current block

## How has this been tested?
Currently only manual test was provided to check if feature listed in the [#1300](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1300) works as expected.
Also still need to fix some failing unit test but I will be cover soon.

## Screenshots 
Please see below screenshot. It presents how to render InspectorControls inside Image block. On left side there is usage on Web. On the right side is the usage on mobile.
Note here that I temporally left `BottomSheet.Cell` to render buttons inside `BottomSheet` - the markup will be also refactor in separate PR to be consistent with web markup.
image

## Types of changes
It should not have impact on usage on web but provides changes in usage on mobile (see comments in `changes` section in this PR)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-editor/src/components/block-list/block.native.js
Comment (jbinda) [inline packages/block-editor/src/components/block-list/block.native.js]: Move `showSettings` and settingsToggle from edit.native.js of image blok to allow open the settings from each block passing onSettingsToggle.

Currently I'm working to move this logic to `InspectorControls` context provider so perhaps it will be possible to do not change `block.native.js`

File: packages/block-editor/src/components/block-list/block.native.js
Comment (jbinda) [inline packages/block-editor/src/components/block-list/block.native.js]: pass to `showSettings` on `onSettingsToggle` to `BlockMobileTolbar` to allow open/hide `BottomSheet` with additional controls to adjust block settings

File: packages/block-editor/src/components/inspector-controls/index.js
Comment (jbinda) [inline packages/block-editor/src/components/inspector-controls/index.js]: Define separate slots for `Buttons` allows to render button to manage the open/close state of the `BottomSheet` and also render the whole content (`Fill` component) of this panel using the same markup: 

`<InspectorControls>
    {content to render here}
</InspectorControls>
`

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [inline packages/block-library/src/image/edit.native.js]: move `BottomSheet` open/close handler to the top (`block.native.js`) as well as `showOptions` prop of the state

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [inline packages/block-library/src/image/edit.native.js]: refactor usage to get markup as closest as possible to web implementation

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [inline packages/block-library/src/image/edit.native.js]: tell Inspector Controls to be render in `InspectorControls.SlotButtons` instead of `InspectorControls.Slot`

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [inline packages/block-library/src/image/edit.native.js]: pass handler from the 'block.native.js` to open/close `BottomSheet` with options

File: packages/block-library/src/video/edit.native.js
Comment (jbinda) [inline packages/block-library/src/video/edit.native.js]: the same as I did in image block

File: packages/components/src/panel/body.native.js
Comment (jbinda) [inline packages/components/src/panel/body.native.js]: prepare simple PanelBody implementation - TBD what do do here next

File: packages/block-library/src/image/edit.native.js
Comment (pinarol) [inline packages/block-library/src/image/edit.native.js]: If I understood correctly you can pass `onSettingsToggle` via fillProps that way you don't need to pass it from component to component. 

I see some example usage in the code base but haven't tested myself:

```
    <__experimentalBlockSettingsMenuFirstItem.Slot
                                        fillProps={ { onClose } }
                           />
```

Passing to the `Fill`:

```
			<__experimentalBlockSettingsMenuFirstItem>
				{ ( { onClose } ) => <BlockInspectorButton onClick={ onClose } /> }
			</__experimentalBlockSettingsMenuFirstItem>
```

**Review by pinarol — COMMENTED** [review]
hey @jbinda Thanks for the PR 🎉 I tried to address all the main things but let me know if I missed something

**Comment by pinarol** [comment]
I am also inviting @koke and @marecar3 as reviewers as `InspectorControls` is one of the most commonly used components.

**Comment by pinarol** [comment]
Let me also link this comment here in case you have different opinions @koke @marecar3 please go ahead and respond here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1300#issuecomment-527127376

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (pinarol) [reply packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: On the second thought, in case we move the settings button here, we won't have anything left to fill this slot. So I'd be OK if we remove InspectorControls.SlotButtons altogether. We can add a new slot here in the future when needed.

**Comment by koke** [comment]
I agree with your [second thought](https://github.com/WordPress/gutenberg/pull/17279#discussion_r319948856). I don't think we'll need the slot for now, and if/when we do it should probably be called something else.

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: Ok I see your point. The settings button should be common and available for all blocks (sometimes it won't be visible because block do not have any settings) and I agree it make sense to move it to `BlockMobileToolbar`.

I will also check what's the option for pass prop to conditionally show/hide it if block has settings

File: packages/components/src/panel/body.native.js
Comment (jbinda) [reply packages/components/src/panel/body.native.js]: Ok, I will exclude PanelBody component and just render View instead.

Maybe if block has more sections the `TabView` after show `BottomSheet` components will be more user-friendly to show more settings sections on mobile ?

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: Another approach is to add prop to `InspectorControls` (e.g `buttons`) to define that this is a slot for buttons:

`<InspectorControls.Slot buttons />`

Then passing info about children in `InspectorControls` Fills to conditionally show/hide `Settings` button should be quite easy 

I will prepare some code to present this solution to make more clear view on that

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: sure !

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: I have also thought about that and already did some work - will present how it works in next commits.

According to add wrapper for BottomSheet.Cell components - that's fine. We will do it separate PR. I have already posted a PR to add RangeCell implementation in #17282

File: packages/components/src/panel/body.native.js
Comment (pinarol) [reply packages/components/src/panel/body.native.js]: sorry if I wasn't clear before, I think we should keep the PanelBody but just render a simple View in it for now.

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [inline packages/block-library/src/image/edit.native.js]: `TollbarButton` which opens the BottomSheet is moved to `BlockMobileTolbar`. It will be visible only when there is some settings wrapped in `InspectorControls` block inside block

File: packages/components/src/panel/body.native.js
Comment (jbinda) [reply packages/components/src/panel/body.native.js]: Done.

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: My last thoughts about that is maybe it's better to open separate PR on that ?

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: condition for hide/show settings is finished

Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: done

File: packages/components/src/slot-fill/index.native.js
Comment (pinarol) [inline packages/components/src/slot-fill/index.native.js]: I didn't understand the need for putting BottomSheet component inside slot-fill. slot-fill is a multi purpose pattern, would it be possible if we don't touch it at all?

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (pinarol) [inline packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: As we [discussed](https://github.com/WordPress/gutenberg/pull/17279#discussion_r319948856) InspectorControls is an independent thing from other toolbar buttons. So it doesn't look right to merge `settings button` with other buttons under **InspectorControls.Slot buttons**. For now we don't have to have a slot for other buttons for simplicity, we can remove BlockMobileToolbarItems altogether and  just add a `ToolbarButton` instead.

File: packages/block-library/src/image/edit.native.js
Comment (pinarol) [reply packages/block-library/src/image/edit.native.js]: yes that's what I said 👍

File: packages/block-editor/src/components/block-list/block.native.js
Comment (pinarol) [reply packages/block-editor/src/components/block-list/block.native.js]: The thing that bugs me is that `showSettings` flag is not used for anything inside this component, it is just passed to `BlockMobileToolbar`. What is more it is toggled by a component outside of this component. So it looks like is doesn't belong here, right?

My [idea](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1300#issuecomment-527127376) was to create a new SettingsToolbarButton component that renders below button and keep `showSettings` in its state but I don't have strong opinions about that.

```
                                               <ToolbarButton
							title={ __( 'Image Settings' ) }
							icon="admin-generic"
							onClick={ onSettingsToggle }
						/>
```

File: packages/components/src/slot-fill/index.native.js
Comment (koke) [reply packages/components/src/slot-fill/index.native.js]: I agree, I don't see the need to change slot-fill for this. I'd recommend looking at how this works on the web:

- `@wordpress/edit-post` has a `Layout` component that includes a `Sidebar.Slot`
- I believe that the sidebar visibility is controlled via the `core/edit-post` store with the `isEditorSidebarOpened` selector. The button to open the sidebar would dispatch the `openGeneralSidebar` action.
- It also includes a `SettingsSidebar` that fills that slot with many things, including `BlockInspector`.
- This `BlockInspector` is where `InspectorControls.Slot` is placed

For mobile, we use a modal `BottomSheet` instead of the `Sidebar` (although I imagine iPads could show a sidebar in the future), and the button to toggle it is in the block toolbar instead of a global editor toolbar.

The UI is a little different but I think the existing web architecture should still fit the native UI.

File: packages/components/src/slot-fill/context.js
Comment (jbinda) [reply packages/components/src/slot-fill/context.js]: I see the point please check my comment above under @koke comment

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: I see the point please check my comment above under @koke comment

File: packages/components/src/slot-fill/index.native.js
Comment (jbinda) [reply packages/components/src/slot-fill/index.native.js]: @koke, @pinarol thank you for review.

I followed with the remarks you have posted and implement changes as below:
1. I have moved `BottomSheet` under `Layout` component (the question here is if we want to wrap it in e.g. <BottomSheet.Slot /> to make it more similar to web or at this stage it will be ok as it is)
2. I use logic from web to show/hide the settings `BottomSheet` container
3. I have created `BlockSettingsButton` and use `slot-fill` logic because I need to conditionally show/hide it depending if given block have some settings wrapped in `InspectorControls`
4 I created native implementation for `InspectorControls` which will render the same as for web plus `SettingsButton`

In current implementation I was able to remove all changes in `slot-fill` logic (only `InspectorControls` for mobile is affected) and also the behaviour is the same as I recorder before. You can noticed that after you delete the children of `InspectorControls` in `edit.native.js` for image block the settings button disappears.

**Review by Tug — COMMENTED** [review]
Nice refactor here 👍 
I like the fact that we removed some of the internal state of the `ImageEdit` component `showSettings` and moved it to the store as `editorSidebarOpened`.
Not sure why we would want a `Slot` for `BlockSettingsButton` though.

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: Thanks !

File: packages/block-editor/src/components/block-settings/container.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-settings/container.native.js]: Ok, I will give it a try and let you know.

File: packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block-mobile-toolbar.native.js]: It's because we render the settings button `Fill` component inside Inspector Controls when it has any children. Using it in that way we can automatically show/hide setting button depending if block has some settings wrappen inside`InspectorControl` component inside `edit.native.js` file.

At this moment it was my best idea to make the API on mobile usage as much as possible to web.

File: packages/block-editor/src/components/block-settings/container.native.js
Comment (jbinda) [reply packages/block-editor/src/components/block-settings/container.native.js]: I have merged mentioned branch and inspector-control branch. It works fine.

The `BottomSheet` comes from `Picker` is nicely show on the top of settings `BottomSheet` and the behaviour looks exactly like in the gif presented by cameronvoell

**Comment by koke** [comment]
> Using it in that way we can automatically show/hide setting button depending if block has some settings wrapped inside InspectorControl component

There was something bugging me about this and I realized that the behavior is different on the web, as the settings button is always present, because the sidebar contains other things as well. After chatting with @iamthomasbishop, we think it would be interesting to have the block settings button always available on every block. I don't think we need to block this PR on that, but it would be good to consider after merging this.

On the web, even if you ignore the document options, block settings always shows the block icon, title, and description, and most blocks have the advanced css field:

Screen Shot 2019-09-23 at 16 57 49

The blocks currently available on mobile don't implement those `InspectorControls` yet in their native variants, but once they do, I'd change the behaviour to match the web.

File: packages/block-editor/src/components/block-settings/container.native.js
Comment (koke) [reply packages/block-editor/src/components/block-settings/container.native.js]: I see that this seems to replace one `BottomSheet` with the new one with the second one is presented. @iamthomasbishop, do you envision more than one sheet being visible at once? I'm not sure we have a use case right now, but looking at how the new iOS 13 modals work, I'd expect to be able to show several levels of bottom sheets stacked on top of each other:

modals

Comment (iamthomasbishop) [reply packages/block-editor/src/components/block-settings/container.native.js]: > do you envision more than one sheet being visible at once?

@koke I think there will probably be cases, and I'd expect it to work similar to the example you gave above. I'm a fan of iOS 13's new native modal sheets, and I'd be even more of a fan if there is a way to present them initially as half-screen sheets, similar to Material modal sheets.

**Comment by iamthomasbishop** [comment]
Agreed w/ @koke – if we add the `additional css class` input on every block, we can safely add (and should add) the `cog`/`settings` trigger to every block. There has been some user confusion around not seeing settings for every block type, and while we don't have _all_ the settings for a bit, it'll at least relieve the confusion around the entry point.

**Comment by jbinda** [comment]
@koke - right! that was the main difference. On web there are others element that has a slot inside Settings Sidebar. On mobile we have less space to show them all at once. Stacked bottom sheets seems to be good solution

@iamthomasbishop - as you mentioned in current implementation not every block has a settings button. So the fastest way to show it in every block is to take care about AdvancedInspectorControl mobile implementation  right ? (this is the component which renders Additional CSS classes)

Summarizing: If it's ok if we merge this PR and then take care of AdvancedInspectorControls to finally refactor settings button show/hide logic ? Or we should wait with merging ?

**Comment by pinarol** [comment]
> Summarizing: If it's ok if we merge this PR and then take care of AdvancedInspectorControls to finally refactor settings button show/hide logic ? Or we should wait with merging ?

I don't think we should block this PR for that, we can handle it separately.

**Comment by jbinda** [comment]
Ok, so I resolve conflicts and it's ready to merge

**Comment by iamthomasbishop** [comment]
Agreed w/ @pinarol – this isn't a blocker. 

> So the fastest way to show it in every block is to take care about AdvancedInspectorControl mobile implementation right ? (this is the component which renders Additional CSS classes)

Yes, essentially just adding the additional css classes input should in theory take care of the problem, bc every block (afaik) has that field.

**Review by pinarol — APPROVED** [review]
LGTM, tested with the WPiOS app. Just added a couple of code cleaning comments. 

@Tug @koke Do you want to give this a second look?

**Comment by pinarol** [comment]
Could you update the branch and resolve the conflicts? @jbinda

**Comment by pinarol** [comment]
Opened an [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1385) to handle advanced settings separately.

**Comment by jbinda** [comment]
Sure !

File: packages/block-library/src/video/edit.native.js
Comment (jbinda) [reply packages/block-library/src/video/edit.native.js]: done!

File: packages/block-library/src/image/edit.native.js
Comment (jbinda) [reply packages/block-library/src/image/edit.native.js]: done!

**Comment by jbinda** [comment]
@pinarol conflicts are solved and Ci passed :)

File: packages/components/src/panel/body.native.js
Comment (koke) [inline packages/components/src/panel/body.native.js]: Minor suggestion, but if this has no custom behavior other than rendering a View, we could just do:

```jsx
/**
  * External dependencies
  */
import { View } from 'react-native';

export default View;
```

File: packages/edit-post/src/components/layout/index.native.js
Comment (koke) [inline packages/edit-post/src/components/layout/index.native.js]: I'm not understanding this. From what I can see, `BottomSheetSettings` is the `Fill` to `BottomSheetSettings.Slot`. If they're going to be rendered next to each other, I don't know why we'd need a slot-fill.

I think this was useful before we used the store to toggle the bottom sheet visibility, but could be simplified now?

File: packages/block-editor/src/components/inspector-controls/index.native.js
Comment (koke) [inline packages/block-editor/src/components/inspector-controls/index.native.js]: I think it would be safer to check `React.Children.count( children ) > 0` rather than see if it's truthy. I can't think of many examples, but if I understand correctly, `children` could be a function (truthy) that returned no children

Comment (jbinda) [reply packages/block-editor/src/components/inspector-controls/index.native.js]: you're right will change that

File: packages/components/src/panel/body.native.js
Comment (jbinda) [reply packages/components/src/panel/body.native.js]: It's because I wrapped all content inside `Inspector Controls` in `PanelBody` 

At this moment I also know that there will be additional work referring to PanelBody stars in few days so would you mind if we can keep it as it is ?

Comment (koke) [reply packages/components/src/panel/body.native.js]: I think it's ok to keep it if it's going to get filled in soon

File: packages/edit-post/src/components/layout/index.native.js
Comment (jbinda) [reply packages/edit-post/src/components/layout/index.native.js]: This pattern comes from web version. The `Sidebar` is constructed in similar way. I've picked up this solution because it do not brings a lot of additional login in `Layout` component but allows to move it over. 

Another reason is the file structure - in current implementation I import `BottomSheetSettings` with its logic from `@wordpress/block-editor` and the layout just render it.

I know that `Sidebar` is much more complex but it seems that our `BottomSheet` settings component also will be growing fast so that's the third reason behind.

However as you noticed we can took whole `BottomSheetFill` and render it in `Layout` component. After pinning some `selectors` it will also work.

You have much more experience and knowledge with this repo so let me know wdyt

File: packages/block-editor/src/components/inspector-controls/index.native.js
Comment (jbinda) [reply packages/block-editor/src/components/inspector-controls/index.native.js]: done

File: packages/edit-post/src/components/layout/index.native.js
Comment (koke) [reply packages/edit-post/src/components/layout/index.native.js]: I think the web part is a bit different because there can be several sidebar fills via plugins. At least it is made clearer by it using [two different names for it](https://github.com/WordPress/gutenberg/blob/a5b1b2e271a55a7474bdd803df490871252a2d39/packages/edit-post/src/components/layout/index.js#L125):

```jsx
<SettingsSidebar />
<Sidebar.Slot />
```

So first it defines the contents of the SettingsSidebar (document + block settings) and then it renders what actually contains the sidebar. It's still a bit awkward to see them side by side but it makes sense because `SettingsSidebar` is more clearly one of possibly several things that will render in the sidebar.

I guess the question is whether we expect to support anything else other than block settings to fill that slot. With the current design, I don't think we do, because if we added a second fill, it would render alongside the block inspector (like in the sidebar), but it would present a second bottom sheet. If we wanted to support other things, we'd have to start using sidebar names (e.g. `edit-post/block`) and make sure we only [present the matching one](https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/sidebar/index.js#L55) instead of just checking for `isEditorSidebarOpened`.

If we only expect to put the InspectorControls of the selected block in it, I still think we could simplify this.

Comment (jbinda) [reply packages/edit-post/src/components/layout/index.native.js]: how about that ?

image

Comment (koke) [reply packages/edit-post/src/components/layout/index.native.js]: I think that looks right 👌

Comment (jbinda) [reply packages/edit-post/src/components/layout/index.native.js]: ok so I will push this in a minutes

**Comment by pinarol** [comment]
@jbinda could you update target branch as master?
@koke could you give a quick re-review afterwards and then we can merge it right?

**Comment by jbinda** [comment]
@pinarol already in-progress :)

**Comment by koke** [comment]
Yes, code looks good. I was planning to do some final testing today before merging

**Comment by jbinda** [comment]
I just wonder if we shouldn't first merge `rmobile/master` to `master` ?

**Comment by pinarol** [comment]
It is already done by https://github.com/WordPress/gutenberg/pull/17228

**Comment by jbinda** [comment]
Ok, thanks !

**Review by koke — APPROVED** [review]
Looking good. I saw one failing test, so I restarted the build on Travis to see if it was a glitch. Ready to merge when green IMO

**Comment by jbinda** [comment]
ok, thanks for restart the test.

It's something with unit test. Locally it also fails
image

**Comment by jbinda** [comment]
And there is also failing test on Android on `develop` branch in `gutenberg-master`. Try investigate what's the cause of failure on this tests.

**Comment by jbinda** [comment]
@koke @pinarol  - it seems that it really was a glitch. I didn't found anything that can break test and after rerun CI it passed in both repo so I think it's ready to merge.

**Review by pinarol — APPROVED** [review]
LGTM 🎉

**Comment by jbinda** [comment]
great ! Thanks

**Pull request merged**
